README.md
# Wavelet Matrix for Rust language

[![Build Status](https://travis-ci.org/sekineh/wavelet-matrix-rs.svg?branch=master)](https://travis-ci.org/sekineh/wavelet-matrix-rs)

It provides the various analytics on very large sequence of unsigned integers in constant time.

## Usage

After adding to Cargo.toml, add this line to lib.rs or main.rs.

```rust
extern crate wavelet_matrix;
```

See crate document top for further examples.

- https://docs.rs/wavelet-matrix/

## Benchmarks

- Overall Performance
  - https://github.com/sekineh/wavelet-matrix-rs/blob/master/BENCH.md
  - Bench is performed on Intel Xeon 2800 MHz

- Error evaluation of O(1) SUM
  - https://github.com/sekineh/wavelet-matrix-rs/blob/master/BENCH_SUM.md

## Features

Given an unsigned integer sequence T, it provides the following queries.

### Basic operations

- `.len()`:
  - Returns the length of `T`.
  - It's almost no overhead, just returning the stored value.
- `.lookup(pos)`:
  - Returns the value at the position of T, `T[pos]`.
  - It's slower than array lookup but still is O(1). You might want to save the original Vector for faster lookup.

### Counting

Counting is performed in O(1).

- `.count(start..end, value)`:
  - Returns the number of the element `e` which satisfies `e == value` included in `T[start..end]`
- `.count_prefix(start..end, value, ignore_bit)`:
  - Returns the number of the element `e` which satisfies `e >> ignore_bit == value >> ignore_bit` included in `T[start..end]`
  - This will be useful for counting the number of IPv4 address that satisfies IPv4 prefix such as `192.168.10.0/24`. In this case, the ignore_bit will be 8.
- `.count_lt(start..end, value)`:
  - Returns the number of the element `e` which satisfies `e < value` included in `T[start..end]`
- `.count_gt(start..end, value)`:
  - Returns the number of the element `e` which satisfies `e > value` included in `T[start..end]`
- `.count_range(start..end, val_start..val_end)`:
  - Returns the number of the element `e` which satisfies `val_start <= e < val_end` included in `T[start..end]`

### Searching

Searching is performed in O(1) per a next index.

- `.search(start..end, value)`:
  - Returns the iterator that find indexes of the element `e` which satisfies `e == value` included in `T[start..end]`
- `.search_prefix(start..end, value, ignore_bit)`:
  - Returns the iterator that find indexes of the element `e` which satisfies `e >> ignore_bit == value >> ignore_bit` included in `T[start..end]`
- [TODO] implement various conditions other than equal.

### Ranking

Ranking is performed in roughly O(1) with regard to the number of elements `n`.

- `.max_k(start..end, val_start..val_end, k)`:
  - list the `(value, count)` pairs in descending order.
  - values are constrained to the range `val_start..val_end`.
- `.min_k(start..end, val_start..val_end, k)`:
  - list the `(value, count)` pairs in ascending order.
  - values are constrained to the range `val_start..val_end`.

`.top_k()` is also performed in O(1) in best case, but may take O(n) in the worst case where every value occurs only once!

- `.top_k(start..end, val_start..val_end, k)`:
  - list the `(value, count)` pairs in most-frequent-one-first order.
  - values are constrained to the range `val_start..val_end`.
  - [TODO] clarify the order of same count.

To achieve O(1) performance regardless of the number of unique values, use `.top_k_ranges()` instead:

- [EXPERIMENTAL] `.top_k_ranges(start..end, val_start..val_end, k)`:
  - list the `(v_start..v_end, count)` pairs in most-frequent-one-first order.
  - unlike `.top_k()`, `.top_k_ranges()` returns the exhaustive range set that covers all of the values.
  - values are constrained to the range `val_start..val_end`.
  - [TODO] clarify the order of same count.

### Statistics

#### O(1) Median / O(1) Quantile

- `.median(start..end)`:
  - Returns the median value from `T[start..end]`.
  - same as `.quantile(start..end, start + (end - start) / 2)`.
- `.quantile(start..end, k)`:
  - Returns the (k+1)th smallest value from `T[start..end]`.

#### O(1) Sum / O(1) Average / O(1) Variance

##### Experiment 1

These methods use `.top_k_ranges()` to enumerate the most relevant value ranges.  

They are not as accurate as the method used in Experiment 2.

- [EXPERIMENTAL] `.sum_experiment1(start..end, val_start..val_end, k)`:
  - Approximately calculate the sum of `T[start..end]` using up to `k` wavelet tree nodes.
  - Only values included in the range `val_start..val_end` are processed.
  - To get the exact result, specify `k = m + 1` where `m` is the number of values which are unique.
- [EXPERIMENTAL] `.mean_experiment1(start..end, val_start..val_end, k)`:
  - Approximately calculate the average of `T[start..end]` using up to `k` wavelet tree nodes.
  - Only values included in the range `val_start..val_end` are processed.
  - To get the exact result, specify `k = m + 1` where `m` is the number of values which are unique.
- [EXPERIMENTAL] `.variance_experiment1(start..end, val_start..val_end, k)`:
  - Approximately calculate the variance of `T[start..end]` using up to `k` wavelet tree nodes.
  - Only values included in the range `val_start..val_end` are processed.
  - To get the exact result, specify `k = m + 1` where `m` is the number of values which are unique.

##### Experiment 2

Improvement over Experiment 1.  They use custom node enumerator to minimize the error.

- [EXPERIMENTAL] `.sum_experiment2(start..end, val_start..val_end, k)`:

##### Experiment 3

Improvement over Experiment 2.  They use `Range<u64>` to tell how accurate the computed value is.

- [EXPERIMENTAL] `.sum_experiment3(start..end, val_start..val_end, k)`:

### Classical WaveletMatrix operations

- `.rank(pos, value)`: counts value included in `T[0..pos]`.
  - Note: pos is exclusive. When pos is 0, `.rank()` always returns 0.
- `.select(rank, value)`: return the position of the (rank+1)-th value
  - Note: When found nothing, it returns `.len()` instead of None.

### TODO

- Add Benchmark.
  - Implement same queries using trivial algorithm
  - Compare wm's queries against trivial one.
  - Make a nice plot.
- Profiling
- Optimize underlying rsdic structure.
- Add travis CI.
- Add u128 support or arbitrary-length integer support

- The fastest implementation on the planet

## Credits

- A Go package for myriad array operations using wavelet trees
  - https://github.com/hillbig/waveletTree
  - Basically, the algorithm is deeply derived from the above Go implementation.

- excellent slides in Japanese
  - https://www.slideshare.net/pfi/ss-15916040
  - https://www.slideshare.net/bonprosoft/the-wavelet-matrix

- The original inventor's pdf:
  - http://www.dcc.uchile.cl/~gnavarro/ps/spire12.4.pdf


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
- [CI] Github action for ubuntu/windows/macos + stable/1.56

### Changed
- minimum Rustc version is now 1.56
- rust edition is now "2018"

### Removed
- [CI] travis build is removed

## v0.4.7

- Fix test bug #4
- Fix test bug #3

## v0.4.6

- Add test for `.median()` and `.quantile()`. 
- Add index range check.

## v0.4.5

- Fix dependent crate.

## v0.4.4

- Add `.median()` and `.quantile()`. They are quite fast, only take 3-5 us on 16-bit values.
- Add `.top_k_ranges()` which is faster than `.top_k()` in worst case.
- Add `.sum_experiment1()`, `.mean_experiment1()` and `.variance_experiment1()`.
- Add `.sum_experiment2()`.
- Add `.sum_experiment3()`.
- Add `BENCH.md` bench report.
- Add `BENCH_SUM.md` bench report.

## v0.4.3

- Add `.bit_len()`.
- Add `.dim()`.
- Move examples to crate's document top.
- Add test for `.top_k()`, `.max_k()` and `min_k()`.
- Suppress warnings.

## v0.4.2

- Add `.top_k()`, `.max_k()` and `min_k()`.

## v0.4.1

- Add `.search()` and `.search_prefix()`.

## v0.4.0

- Add `.count()`, `.count_prefix()`, `.count_lt()`, `.count_gt()` and `.count_range()`.
- [INCOMPATIBLE] WaveletMatrix::new() takes `&Vec<u64>`, instead of `Vec<u64>`

## v0.3.0

- [INCOMPATIBLE] .select() now returns .len() instead of None.


src/lib.rs
//! This crate provides state-of-the-art O(1) or alike queries on large number of unsigned integers.
//!
//! The typical memory usage of this data structure is calculated like below.
//!
//! - `bit_len * num_of_elements / 8 * 1.25` [bytes]
//!
//! In other words, it roughly consumes additional 25% space compared with the original data.
//!
//! # Example
//!
//! ```rust
//! use wavelet_matrix::WaveletMatrix;
//!
//! let vec: Vec<u64> = vec![1, 2, 4, 5, 1, 0, 4, 6, 2, 9, 2, 0];
//! //                       0  1  2  3  4  5  6  7  8  9 10 11 (length = 12)
//! let wm = WaveletMatrix::new(&vec);
//!
//! // Basic access
//! assert_eq!(wm.len(), vec.len());
//! assert_eq!(wm.lookup(7), vec[7]);
//! assert_eq!(wm.dim(), 10); // max value + 1
//! assert_eq!(wm.bit_len(), 4); // bit length stored internally
//!
//! // Counting
//! assert_eq!(wm.count(0..wm.len(), 2), 3);
//! assert_eq!(wm.count(0..wm.len(), 4), 2);
//! assert_eq!(wm.count(0..wm.len(), 5), 1);
//! assert_eq!(wm.count(0..wm.len(), 7), 0);
//! assert_eq!(wm.count(0..wm.len(), 39), 0);
//!
//! assert_eq!(wm.count_prefix(0..wm.len(), 8, 3), 1);
//! assert_eq!(wm.count_prefix(0..wm.len(), 6, 1), 1);
//! assert_eq!(wm.count_prefix(0..wm.len(), 0, 1), 4);
//! assert_eq!(wm.count_prefix(0..wm.len(), 0, 2), 7);
//!
//! assert_eq!(wm.count_lt(0..wm.len(), 2), 4);
//! assert_eq!(wm.count_lt(0..wm.len(), 7), 11);
//!
//! assert_eq!(wm.count_gt(0..wm.len(), 2), 5);
//! assert_eq!(wm.count_gt(0..wm.len(), 7), 1);
//!
//! assert_eq!(wm.count_range(0..wm.len(), 0..wm.dim()), 12);
//! assert_eq!(wm.count_range(0..wm.len(), 4..6), 3);
//!
//! // Searching
//! assert_eq!(wm.search(0..wm.len(), 4).collect::<Vec<usize>>(), vec![2, 6]);
//! assert_eq!(wm.search(3..wm.len(), 4).collect::<Vec<usize>>(), vec![6]);
//! assert_eq!(wm.search(0..wm.len(), 7).collect::<Vec<usize>>(), vec![]);
//!
//! // Ranking: (value, count), frequent values first
//! assert_eq!(wm.top_k(0..wm.len(), 0..wm.dim(), 12),
//!            vec![(2, 3), (1, 2), (4, 2), (0, 2), (5, 1), (6, 1), (9, 1)]);
//! assert_eq!(wm.top_k(0..wm.len(), 0..wm.dim(), 4),
//!            vec![(2, 3), (1, 2), (4, 2), (0, 2)]);
//! assert_eq!(wm.top_k(0..wm.len(), 2..9, 12),
//!            vec![(2, 3), (4, 2), (5, 1), (6, 1)]);
//!
//! // Ranking: (value, count), max values first
//! assert_eq!(wm.max_k(0..wm.len(), 0..wm.dim(), 12),
//!            vec![(9, 1), (6, 1), (5, 1), (4, 2), (2, 3), (1, 2), (0, 2)]);
//! assert_eq!(wm.max_k(0..wm.len(), 0..wm.dim(), 4),
//!            vec![(9, 1), (6, 1), (5, 1), (4, 2)]);
//! assert_eq!(wm.max_k(0..wm.len(), 2..9, 12),
//!            vec![(6, 1), (5, 1), (4, 2), (2, 3)]);
//!
//! // Ranking: (value, count), min values first
//! assert_eq!(wm.min_k(0..wm.len(), 0..wm.dim(), 12),
//!            vec![(0, 2), (1, 2), (2, 3), (4, 2), (5, 1), (6, 1), (9, 1)]);
//! assert_eq!(wm.min_k(0..wm.len(), 0..wm.dim(), 4),
//!            vec![(0, 2), (1, 2), (2, 3), (4, 2)]);
//! assert_eq!(wm.min_k(0..wm.len(), 2..9, 12),
//!            vec![(2, 3), (4, 2), (5, 1), (6, 1)]);
//!
//! // Statistics
//! assert_eq!(wm.quantile(0..wm.len(), 0), 0);
//! assert_eq!(wm.quantile(0..wm.len(), 8), 4);
//! assert_eq!(wm.quantile(0..wm.len(), 11), 9);
//! ```

extern crate succinct;

mod node_range;
mod rsdic_simple;
mod wavelet_matrix;

pub use crate::wavelet_matrix::*;


src/node_range.rs
use std::cmp::Ordering;
use std::ops::Range;

/// Stores the node position during Wavelet Matrix traversing.
#[derive(Debug, Clone, Eq, PartialEq, PartialOrd)]
pub struct QueryOnNode {
    pub pos_start: usize,
    pub pos_end: usize,
    pub prefix_char: u64,
    pub depth: u8,
    pub bit_len: u8,
}

impl QueryOnNode {
    pub fn new(pos_range: Range<usize>, prefix_char: u64, depth: u8, bit_len: u8) -> Self {
        QueryOnNode {
            pos_start: pos_range.start,
            pos_end: pos_range.end,
            prefix_char: prefix_char,
            depth: depth,
            bit_len: bit_len,
        }
    }
    #[inline]
    fn count(&self) -> usize {
        self.pos_end - self.pos_start
    }
    #[inline]
    fn value_range(&self) -> Range<u64> {
        let shift = self.bit_len - self.depth;
        self.prefix_char << shift..(self.prefix_char + 1) << shift
    }
    #[inline]
    fn sum_error(&self) -> u64 {
        let vr = self.value_range();
        (vr.end - vr.start - 1) * (self.count() as u64)
    }
    // #[inline]
    // fn sum_expected(&self) -> u64 {
    //     let vr = self.value_range();
    //     (vr.end + vr.start) / 2 * (self.count() as u64)
    // }
}

pub fn cmp_by_frequency(self_: &QueryOnNode, other: &QueryOnNode) -> Ordering {
    if self_.count() != other.count() {
        self_.count().cmp(&other.count()) // wide node first
    } else if self_.depth != other.depth {
        self_.depth.cmp(&other.depth) // deepest node first
    } else {
        self_.pos_start.cmp(&other.pos_start) // just to make the order more predictable
    }
}

pub fn cmp_by_sum_error(self_: &QueryOnNode, other: &QueryOnNode) -> Ordering {
    self_.sum_error().cmp(&other.sum_error())
}

pub fn cmp_by_descending(self_: &QueryOnNode, other: &QueryOnNode) -> Ordering {
    self_.prefix_char.cmp(&other.prefix_char)
}

pub fn cmp_by_ascending(self_: &QueryOnNode, other: &QueryOnNode) -> Ordering {
    other.prefix_char.cmp(&self_.prefix_char)
}


benches/bencher.rs
#![allow(non_snake_case)]
// #[macro_use]
// extern crate bencher;
#![feature(test)]
extern crate rand;
extern crate test;
extern crate wavelet_matrix;

// use bencher::Bencher;
use test::Bencher;
use wavelet_matrix::WaveletMatrix;
use rand::distributions::IndependentSample;

fn new_helper(b: &mut Bencher, num: usize, upper: u64, bytes_per_elem: u64) {
    let mut rng = rand::weak_rng();
    let range = rand::distributions::Range::new(0, upper);
    let vec: Vec<u64> = std::iter::repeat(0)
        .take(num)
        .map(|_| range.ind_sample(&mut rng) as u64)
        .collect::<Vec<_>>();

    b.iter(|| WaveletMatrix::new(&vec));
}

/// bench new() with various bit length
mod new_by_bits {
    use ::*;

    #[bench]
    fn new_8bit_10000(b: &mut Bencher) {
        new_helper(b, 10000, std::u8::MAX as u64, 1)
    }

    #[bench]
    fn new_16bit_10000(b: &mut Bencher) {
        new_helper(b, 10000, std::u16::MAX as u64, 2)
    }

    #[bench]
    fn new_32bit_10000(b: &mut Bencher) {
        new_helper(b, 10000, std::u32::MAX as u64, 4)
    }

    #[bench]
    fn new_64bit_10000(b: &mut Bencher) {
        new_helper(b, 10000, std::u64::MAX as u64, 8)
    }
}

/// bench new() with various number of elements
mod new_by_N {
    use ::*;

    #[bench]
    fn new_16bit____1000(b: &mut Bencher) {
        new_helper(b, 1000, std::u16::MAX as u64, 2)
    }

    #[bench]
    fn new_16bit___10000(b: &mut Bencher) {
        new_helper(b, 10000, std::u16::MAX as u64, 2)
    }

    #[bench]
    fn new_16bit__100000(b: &mut Bencher) {
        new_helper(b, 100000, std::u16::MAX as u64, 2)
    }

    // #[bench]
    fn new_16bit_1000000(b: &mut Bencher) {
        new_helper(b, 1000000, std::u16::MAX as u64, 2)
    }
}

use rand::Rng;

fn lookup_helper(b: &mut Bencher, num: usize, upper: u64, bytes_per_elem: u64) {
    let mut rng = rand::weak_rng();
    let range = rand::distributions::Range::new(0, upper);
    let vec: Vec<u64> = std::iter::repeat(0)
        .take(num)
        .map(|_| range.ind_sample(&mut rng) as u64)
        .collect::<Vec<_>>();

    let wm = WaveletMatrix::new(&vec);

    b.iter(|| {
        let idx = rng.next_u64() as usize % num;
        assert_eq!(wm.lookup(idx), vec[idx]); // vec[idx] is negrigible
    });
}

mod lookup {
    use ::*;

    #[bench]
    fn lookup_32bits____10000(b: &mut Bencher) {
        lookup_helper(b, 10000, std::u32::MAX as u64, 4);
    }

    #[bench]
    fn lookup_32bits_10000000(b: &mut Bencher) {
        lookup_helper(b, 10000000, std::u32::MAX as u64, 4);
    }

    #[bench]
    fn lookup_64bits____10000(b: &mut Bencher) {
        lookup_helper(b, 10000, std::u64::MAX as u64, 8);
    }

    #[bench]
    fn lookup_64bits_10000000(b: &mut Bencher) {
        lookup_helper(b, 10000000, std::u64::MAX as u64, 8);
    }
}


benches/easybench.rs
extern crate easybench;
use easybench::{bench, bench_env};
use rand::Rng;
extern crate rand;
extern crate wavelet_matrix;

use rand::distributions::IndependentSample;
use wavelet_matrix::WaveletMatrix;

struct WaitCooldown {
    fastest: f64,
    wait: f64,
    ratio: f64,
    backoff: f64,
}

impl WaitCooldown {
    fn new() -> WaitCooldown {
        let result = WaitCooldown::measure();
        let time = result.ns_per_iter;
        WaitCooldown {
            fastest: time as f64,
            wait: 1.0,
            ratio: 1.1,
            backoff: 1.2,
        }
    }

    fn wait(&mut self) {
        loop {
            let result = WaitCooldown::measure();
            let time = result.ns_per_iter;
            if time > self.fastest * self.ratio {
                let dur = std::time::Duration::from_secs(self.wait as u64);
                std::thread::sleep(dur);
                self.wait *= self.backoff;
            } else if time < self.fastest {
                self.fastest = time;
                break;
            } else {
                break;
            }
        }

        self.wait = 1.0;
    }

    fn measure() -> easybench::Stats {
        let rng = rand::weak_rng();
        let num = std::u64::MAX;
        let stats = bench_env(rng.clone(), |rng| rng.gen_range(0, num));
        println!("(wait for cpu cooldown) {}", stats);
        stats
    }
}

fn overall_helper(num: usize, desc: &str, upper: u64, _limit_secs: u64) {
    let mut rng = rand::weak_rng();
    let range = rand::distributions::Range::new(0, upper);
    let vec: Vec<u64> = std::iter::repeat(0)
        .take(num)
        .map(|_| range.ind_sample(&mut rng) as u64)
        .collect::<Vec<_>>();

    println!("### N = {}", num);
    println!("");
    println!("```");

    use std::time::*;
    let time_start = Instant::now();
    let wm = WaveletMatrix::new(&vec);
    let dur = time_start.elapsed();
    let dur_ns = dur.as_secs() as f64 * 1e9 + dur.subsec_nanos() as f64;
    println!(
        "{:>24}, N = {}, {}: {:>10} ns   (only 1 iteration)",
        "WaveletMatrix::new()", num, desc, dur_ns,
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".lookup(rand)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let idx = rng.gen_range(0, num);
            wm.lookup(idx)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        "vec[rand]",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let idx = rng.gen_range(0, num);
            vec[idx]
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        "rand only",
        num,
        desc,
        bench_env(rng.clone(), |rng| rng.gen_range(0, num))
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".rank()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let pos = rng.gen_range(0, num);
            let value = rng.gen_range(0, wm.dim());
            wm.rank(pos, value)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".select()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let rank = rng.gen_range(0, num);
            let value = rng.gen_range(0, wm.dim());
            wm.select(rank, value)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".select_lt()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let rank = rng.gen_range(0, num);
            let value = rng.gen_range(0, wm.dim());
            wm.select_lt(rank, value)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".count()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let value = rng.gen_range(0, wm.dim());
            wm.count(start..end, value)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".count_prefix()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let value = rng.gen_range(0, wm.dim());
            wm.count_prefix(start..end, value, wm.bit_len() / 2)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".search().next()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let value = rng.gen_range(0, wm.dim());
            wm.search(start..end, value).next()
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".search_prefix().next()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let value = rng.gen_range(0, wm.dim());
            wm.search_prefix(start..end, value, wm.bit_len() / 2).next()
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".median()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            wm.median(start..end)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".quantile()",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let k = rng.gen_range(0, end - start);
            wm.quantile(start..end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".top_k(k=16)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 16;
            wm.top_k(start..end, val_start..val_end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".top_k_ranges(k=16)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 16;
            wm.top_k_ranges(start..end, val_start..val_end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".top_k_ranges(k=256)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 256;
            wm.top_k_ranges(start..end, val_start..val_end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".top_k_ranges(k=65536)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 65536;
            wm.top_k_ranges(start..end, val_start..val_end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".sum_ex3(k=16)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 16;
            wm.sum_experiment3(start..end, val_start..val_end, k)
        })
    );
    println!(
        "{:>24}, N = {}, {}: {}",
        ".sum_ex3(k=256)",
        num,
        desc,
        bench_env(rng.clone(), |rng| {
            let end = rng.gen_range(0, num);
            let start = rng.gen_range(0, end);
            let val_end = rng.gen_range(0, wm.dim());
            let val_start = rng.gen_range(0, val_end);
            let k = 256;
            wm.sum_experiment3(start..end, val_start..val_end, k)
        })
    );
    println!("```");
    println!("");
}

fn overall_performance() {
    let long_test = true;

    let mut wcd = WaitCooldown::new();

    println!("# Overall Performance");
    println!("");

    println!("## 16-bit Values");
    println!("");

    if long_test {
        wcd.wait();
        overall_helper(10000000, "16-bit values", std::u16::MAX as u64, 50);
        wcd.wait();
        overall_helper(1000000, "16-bit values", std::u16::MAX as u64, 5);
    }
    wcd.wait();
    overall_helper(100000, "16-bit values", std::u16::MAX as u64, 1);
    wcd.wait();
    overall_helper(10000, "16-bit values", std::u16::MAX as u64, 1);
    wcd.wait();
    overall_helper(1000, "16-bit values", std::u16::MAX as u64, 1);

    println!("## 32-bit Values");
    println!();

    if long_test {
        wcd.wait();
        overall_helper(10000000, "32-bit values", std::u32::MAX as u64, 100);
        wcd.wait();
        overall_helper(1000000, "32-bit values", std::u32::MAX as u64, 10);
    }
    wcd.wait();
    overall_helper(100000, "32-bit values", std::u32::MAX as u64, 3);
    wcd.wait();
    overall_helper(10000, "32-bit values", std::u32::MAX as u64, 1);
    wcd.wait();
    overall_helper(1000, "32-bit values", std::u32::MAX as u64, 1);

    println!("## 64-bit Values");
    println!();

    if long_test {
        wcd.wait();
        overall_helper(10000000, "64-bit values", std::u64::MAX as u64, 200);
        wcd.wait();
        overall_helper(1000000, "64-bit values", std::u64::MAX as u64, 20);
    }
    wcd.wait();
    overall_helper(100000, "64-bit values", std::u64::MAX as u64, 6);
    wcd.wait();
    overall_helper(10000, "64-bit values", std::u64::MAX as u64, 1);
    wcd.wait();
    overall_helper(1000, "64-bit values", std::u64::MAX as u64, 1);
}

fn statstical_helper(num: usize, lower: u64, upper: u64, k: usize) {
    println!("```");

    let desc = format!("{} <= val < {}", lower, upper);
    let mut rng = rand::weak_rng();
    let range = rand::distributions::Range::new(lower, upper);
    let vec: Vec<u64> = std::iter::repeat(0)
        .take(num)
        .map(|_| range.ind_sample(&mut rng) as u64)
        .collect::<Vec<_>>();
    let wm = WaveletMatrix::new(&vec);

    if wm.dim() < std::u32::MAX as u64 {
        // let k = 5;

        let actual = vec.iter().sum::<u64>();
        println!(
            "{:>24}, N = {}, {}: {:?}",
            "actual sum",
            num,
            desc,
            actual,
            // bench(|| vec.iter().sum::<u64>())
        );

        let computed = wm.sum_experiment1(0..wm.len(), 0..wm.dim(), k);
        let computed_sum = computed.0;
        println!(
            "{:>24}, N = {}, {}: {:?}, error = {}%, {}",
            format!(".sum_expreriment1(k={})", k),
            num,
            desc,
            computed,
            error_pct(computed_sum, actual),
            bench(|| wm.sum_experiment1(0..wm.len(), 0..wm.dim(), k)),
        );

        let computed = wm.sum_experiment2(0..wm.len(), 0..wm.dim(), k);
        let computed_sum = computed.0;
        println!(
            "{:>24}, N = {}, {}: {:?}, error = {}%, {}",
            format!(".sum_expreriment2(k={})", k),
            num,
            desc,
            computed,
            error_pct(computed_sum, actual),
            bench(|| wm.sum_experiment2(0..wm.len(), 0..wm.dim(), k)),
        );

        let computed = wm.sum_experiment3(0..wm.len(), 0..wm.dim(), k);
        let computed_sum = average(&computed.0);
        println!(
            "{:>24}, N = {}, {}: {:?}, error = {}%, {}",
            format!(".sum_expreriment3(k={})", k),
            num,
            desc,
            computed,
            error_pct(computed_sum, actual),
            bench(|| wm.sum_experiment3(0..wm.len(), 0..wm.dim(), k)),
        );
    }
    println!("```");
    println!("");
}

/// perform test and find k for `accuracy_pct`
fn sum_uniform_find_sufficient_k(num: usize, lower: u64, upper: u64, accuracy_pct: f64) -> usize {
    println!("```");
    let mut errorpct = 0.0;

    let desc = format!("{} <= val < {}", lower, upper);
    let mut rng = rand::weak_rng();
    let range = rand::distributions::Range::new(lower, upper);
    let vec: Vec<u64> = std::iter::repeat(0)
        .take(num)
        .map(|_| range.ind_sample(&mut rng) as u64)
        .collect::<Vec<_>>();
    let wm = WaveletMatrix::new(&vec);

    let actual = vec.iter().sum::<u64>();
    println!("{:>24}, N = {}, {}: {:?}", "actual sum", num, desc, actual,);

    let mut k = 1;
    for _ in 0..100 {
        let computed = wm.sum_experiment3(0..wm.len(), 0..wm.dim(), k);
        let computed_sum = average(&computed.0);
        errorpct = error_pct(computed_sum, actual);
        println!(
            "{:>24}, N = {}, {}: {:?}, error = {}%, {}",
            format!(".sum_expreriment3(k={})", k),
            num,
            desc,
            computed,
            errorpct,
            "(bench omitted)" //bench(|| wm.sum_experiment3(0..wm.len(), 0..wm.dim(), k)),
        );
        if errorpct.abs() < accuracy_pct {
            break;
        } else {
            k += 1;
        }
    }

    println!("```");
    println!("");

    k
}

fn average(r: &std::ops::Range<u64>) -> u64 {
    (r.start + r.end) / 2
}

fn error_pct(computed: u64, actual: u64) -> f64 {
    let error = computed as f64 - actual as f64;
    error / (actual as f64) * 100.0
}

// use rand::distributions::range::SampleRange;

// fn random_upto<T: PartialOrd + SampleRange>(upper: u64) -> u64 {
//     rand::weak_rng().gen_range(0, upper)
// }

// fn random_range<T: PartialOrd + SampleRange>(lower: T, upper: T) -> T {
//     rand::weak_rng().gen_range(lower, upper)
// }

fn statistical_performance() {
    let num = 10000;
    println!("# Statistical Performance");
    println!("");
    println!("## Uniform distribution");
    println!("");
    println!(
        "For uniform distribution, `.sum()` produce good results using smaller number of \
              `k`."
    );
    println!("");

    statstical_helper(num, 0, 256, 5);
    statstical_helper(num, 0, 257, 5);
    statstical_helper(num, 256, 1024, 5);
    statstical_helper(num, 255, 1024, 5);
    statstical_helper(num, 256, 1025, 5);

    statstical_helper(num, 400, 2000, 5);

    println!("### Some value ranges require greater k to achieve error < 1%");
    println!("");
    println!("#### Bad (k=5)");
    println!("");
    statstical_helper(num, 300, 1500, 5);

    println!("#### Good (k=10)");
    println!("");
    statstical_helper(num, 300, 1500, 10);
}

fn statistical_sufficient_k() {
    println!();
    println!("### Finding sufficient k that achieve error < 1% for various ranges");
    println!();

    let num = 1000;
    let trial = 100;
    let mut rng = rand::weak_rng();
    let mut ks = std::collections::BTreeMap::new();

    for _ in 0..trial {
        let a = rng.gen_range(0, 65536);
        let b = rng.gen_range(0, 65536);
        let lower = std::cmp::min(a, b);
        let upper = std::cmp::max(a, b);

        let k = sum_uniform_find_sufficient_k(num, lower, upper, 1.0);

        let k_entry = ks.entry(k).or_insert(0);
        *k_entry += 1;
    }
    println!("The distribution of sufficient k is:");
    println!();
    println!("```");
    for k in ks {
        println!(
            "{:>2}: {:>2} {}",
            k.0,
            k.1,
            std::iter::repeat('*').take(k.1).collect::<String>()
        );
    }
    println!("```");
    println!();
}

fn main() {
    // BENCH.md
    overall_performance();

    // BENCH_SUM.md
    // statistical_performance();
    // statistical_sufficient_k();
}
